### PR TITLE
Weighted endpoints RPCClient

### PR DESCRIPTION
**Pull request by cryptogosu**
Added ability to have a weighted switch between different urls.

### CODE REVIEW

File: src/main/java/org/p2p/solanaj/rpc/WeightedCluster.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package org.p2p.solanaj.rpc;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import org.p2p.solanaj.rpc.types.WeightedEndpoint;
+
+import java.util.List;
+
+@Data
```
Comment (busligisti) [inline src/main/java/org/p2p/solanaj/rpc/WeightedCluster.java:10]: Are all the methods generated by `@Data` necessary here? Only the getters and the all args constructor are used so far. Please consider this in the WeightedEndpoint class too.

Comment (skynetcapital) [reply src/main/java/org/p2p/solanaj/rpc/WeightedCluster.java:10]: > Are all the methods generated by `@Data` necessary here? Only the getters and the all args constructor are used so far. Please consider this in the WeightedEndpoint class too.

+1, will merge when addressed + the small docs

**Review by skynetcapital — CHANGES_REQUESTED** [review]
See comment by @busligisti on potential requested change, plus the docs mentioned in Discord.

File: src/main/java/org/p2p/solanaj/rpc/RpcClient.java [hunk 2]
```diff
@@ -73,7 +79,27 @@ public RpcApi getApi() {
...
 
+    private String getWeightedEndpoint() {
+        int currentNumber = 0;
+        int randomMultiplier = cluster.endpoints.stream().mapToInt(WeightedEndpoint::getWeight).sum();
+        double randomNumber = Math.random() * randomMultiplier;
+        String currentEndpoint = "";
+        for (WeightedEndpoint endpoint: cluster.endpoints) {
+            if (randomNumber > currentNumber + endpoint.getWeight()) {
+                currentNumber += endpoint.getWeight();
+            } else if (randomNumber >= currentNumber && randomNumber < currentNumber + endpoint.getWeight()) {
```
Comment (busligisti) [inline src/main/java/org/p2p/solanaj/rpc/RpcClient.java:96]: I just noticed this one. Do we need this else-if branch? The same code is executed inside both this and the else branch.

Comment (cryptogosu) [reply src/main/java/org/p2p/solanaj/rpc/RpcClient.java]: Updated my else if to use `randomNumber <= currentNumber + endpoint.getWeight()` to remove the last `else`

File: src/main/java/org/p2p/solanaj/rpc/WeightedCluster.java (hunk 1 above)
Comment (cryptogosu) [reply src/main/java/org/p2p/solanaj/rpc/WeightedCluster.java]: Updated to use `@Getter`, `@Setter`, `@AllArgsConstructor`, and `@NoArgsConstructor`. Added both the last two to give more flexibility to developers on how they create objects.

Usually `@Data` is recommended if the state of the object may change in the future. I expect these objects will be static though

**Comment by skynetcapital** [comment]
LGTM, thanks!

**Pull request merged**